### PR TITLE
[v1.7] Refactor, tests & new accessories

### PR DESCRIPTION
**Pull request by ptz0n**
It's been a while, but we are finally getting ready to release this rewrite covering:

- Make use of JavaScript classes.
- Split accessories into different files.
- Setup code standards to help maintain code style and quality.
- Cover logic with unit tests.
- Added support for door/window sensors and alarm. 🚪 🔒 🏠 🚨

## Please help test alarm & smart lock

It's pretty straight forward, same config with the additional keys `alarmCode` and `doorCode`:

```json
{
  "platform" : "verisure",
  "name" : "Verisure",
  "email": "your@email.com",
  "password": "yourT0p5ecre7Passw0rd",
  "alarmCode": "0000",
  "doorCode": "000000"
}
```

__Update the plugin globally__

```bash
$ npm install -g git://github.com/ptz0n/homebridge-verisure#dev
```

__Alt. run it manually__

```bash
$ git clone https://github.com/ptz0n/homebridge-verisure.git
$ cd homebridge-verisure
$ git checkout dev
$ npm install
$ homebridge -P .
```

See: https://github.com/ptz0n/homebridge-verisure/pull/17, #28

### CODE REVIEW

**Comment by bjornand1** [comment]
Thanks for updating! Bit of a slowpoke, here, I run Homebridge in docker on a Synology. I can install the new version, but then Homebridge fails to run. Should I uninstall the old version first?

**Comment by ptz0n** [comment]
@bjornand1 Thanks for helping out! There seem to be some problems with initiating the alarm accessory. Would be amazing if you could share the homebridge log output.

**Comment by bjornand1** [comment]
Hi, sure I'll help. How can I send you the log?

**Comment by hoefs** [comment]
Hi, it is working for me, but if I push
img_5289
 the "night" button in home, the homebridge service crashes. "Away" and "Home" are working.

**Comment by ptz0n** [comment]
@bjornand1 If you don't want to share it openly here, please send i to erik@eng.se.

@hoefs Thanks for reporting! We did not handle all requested states correctly. This have now been fixed, please update and try again.

**Comment by hoefs** [comment]
@ptz0n should I run this again? npm install -g git://github.com/ptz0n/homebridge-verisure#dev

**Comment by ptz0n** [comment]
@hoefs Yes, that should do it!

**Comment by hoefs** [comment]
@ptz0n what status will the home app show when I arm the alarm "home" as I assume that "home" and "night" is the same for Verisure?

File: accessories/alarm.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    this.alarmCode = alarmCode.toString();
+    this.model = 'ALARM';
+    this.name = VerisureAccessory.getUniqueAccessoryName(`Alarm (${this.installation.config.alias})`);
+
+    const { SecuritySystemCurrentState } = this.homebridge.hap.Characteristic;
+    this.armStateMap = {
+      ARMED_AWAY: [SecuritySystemCurrentState.AWAY_ARM],
+      ARMED_HOME: [
+        SecuritySystemCurrentState.STAY_ARM,
+        SecuritySystemCurrentState.NIGHT_ARM,
```
Comment (ptz0n) [inline accessories/alarm.js]: @hoefs Correct, both HomeKit arm states "Home" & "Night" are both mapped to the Verisure "Home" state.

**Comment by hoefs** [comment]
@ptz0n I have updated yesterday and was able to test this morning. Both "Home" and "Night" result in the Verisure "Home" state and don't crash homebridge.
I still have 1 issue with the arming and disarming (see screenshots): When I disarm (or arm) the alarm via home.app the action is shown (arming/disarming) but the state for the alarm always ends up in "No Response"; only after closing and  reopening the home.app the exact state of the alarm is shown.
1) push "Off"
2) "Disarming is shown
3) "No Response" is shown for several minutes 
4) close app (= push home button = not force quit)
5) open home.app
6) correct state is shown = "Off"

img_5294
img_5295
img_5296
img_5297

**Comment by ptz0n** [comment]
@hoefs Thanks for testing! The issue you are experiencing should be solved by the code pushed this morning. Please update to the latest code & try again 👍

**Comment by hoefs** [comment]
@ptz0n Indeed! Seems to be working now. The state is updated instantly.

Anything else you need me to test?

**Comment by ptz0n** [comment]
@hoefs We need some smart lock testers. Sadly, I don't have one. @kodsnutten?

**Comment by bjornand1** [comment]
I have a smart lock so I can test, but so far I haven't been able to get 1.6 to run

**Comment by hoefs** [comment]
@bjornand1 show your error and maybe we can help. I had to delete the homebridge-verisure-ts package and install this one.

**Comment by hoefs** [comment]
@bjornand1 maybe your config.json is wrong

Here is mine in "platforms":
`"platforms": [
  {
    "platform" : "verisure",
    "name" : "Verisure",
    "email": "my_email@gmail.com",
    "password": "mypassword",
    "alarmCode": "0000"
  }`

**Comment by bjornand1** [comment]
Hmm could that be the issue? Per what I read, I have the last line as doorcode, not alarmcode.

**Comment by hoefs** [comment]
@bjornand1  I Don't have a door lock (not available in Belgium). 

@ptz0n maybe an issue with the door lock implementation? because of the capital C in doorCode or is it not case sensitive?

**Comment by bjornand1** [comment]
Yes, I do have a Yale Doorman

**Comment by bjornand1** [comment]
And I have now tested various versions of alarmcode/doorcode, with/without capital letters, but 1.6 simply stops the restart of homebridge

**Comment by hoefs** [comment]
@bjornand1 to see the logs if it runs as service on raspbian/debian:
sudo journalctl --unit=homebridge.service --since=2018-05-29

Can you also run: sudo npm list -g --depth=0

**Comment by bjornand1** [comment]
OK, I tried "doorCode", but still the same result. Here is the log (I should mention I run Homebrigde on a Synology):

[2018-5-29 19:38:16] Loaded plugin: homebridge-harmonyhub
[2018-5-29 19:38:17] Registering platform 'homebridge-harmonyhub.HarmonyHub'
[2018-5-29 19:38:17] ---
[2018-5-29 19:38:18] Loaded plugin: homebridge-netatmo-schmittx
[2018-5-29 19:38:18] Registering platform 'homebridge-netatmo.Netatmo'
[2018-5-29 19:38:18] ---
[2018-5-29 19:38:18] Loaded plugin: homebridge-platform-ring-video-doorbell
[2018-5-29 19:38:18] Registering platform 'homebridge-platform-ring-video-doorbell.ring-video-doorbell'
[2018-5-29 19:38:18] ---
[2018-5-29 19:38:18] Loaded plugin: homebridge-sensibo-sky
[2018-5-29 19:38:18] Registering platform 'homebridge-sensibo-sky.SensiboSky'
[2018-5-29 19:38:18] ---
[2018-5-29 19:38:18] Loaded plugin: homebridge-verisure
[2018-5-29 19:38:18] Registering platform 'homebridge-verisure.verisure'
[2018-5-29 19:38:18] ---
[2018-5-29 19:38:18] Loaded plugin: homebridge-config-ui-x
[2018-5-29 19:38:18] Registering platform 'homebridge-config-ui-x.config'
[2018-5-29 19:38:18] ---
[2018-5-29 19:38:18] Loaded config.json with 0 accessories and 5 platforms.
[2018-5-29 19:38:18] ---
[2018-5-29 19:38:18] Loading 5 platforms...
[2018-5-29 19:38:18] [Sensibo Sky] Initializing SensiboSky platform...
[2018-5-29 19:38:18] [Sensibo Sky] Fetching Sensibo devices...
[2018-5-29 19:38:18] [Harmony Hub] Initializing HarmonyHub platform...
[2018-5-29 19:38:18] [Verisure] Initializing verisure platform...
[2018-5-29 19:38:18] [Doorbell] Initializing ring-video-doorbell platform...
[2018-5-29 19:38:18] [Netatmo Platform] Initializing Netatmo platform...
[2018-5-29 19:38:19] Loading 0 accessories...
Load homebridge-verisure.verisure
[2018-5-29 19:38:19] [Doorbell] configureAccessory { UUID: '94084e1c-ba2c-42e9-8d30-44ddfaa06237',
  displayName: 'Ytterdør' }
[2018-5-29 19:38:19] [Doorbell] didFinishLaunching
[2018-5-29 19:38:19] [Sensibo Sky] Varmepumpe : AI State:  true , RefreshCycle:  30000 , fixedState, AI, hideFan : auto true false
[2018-5-29 19:38:19] [Sensibo Sky] Device Added (Name: Varmepumpe, ID: JSXfH3N4, Group: pods)
[2018-5-29 19:38:19] [Sensibo Sky] Initializing platform accessory 'Varmepumpe'...
[2018-5-29 19:38:19] [Netatmo Platform] Initializing platform accessory 'Netatmo Bergly Innemodul Skjenk'...
[2018-5-29 19:38:19] [Netatmo Platform] Initializing platform accessory 'Netatmo Bergly Utendørs'...
[2018-5-29 19:38:20] [Harmony Hub] Fetching Logitech Harmony activities...
[2018-5-29 19:38:20] [Verisure] Initializing platform accessory 'Rökdetektor (Gang nede)'...
[2018-5-29 19:38:20] [Verisure] Initializing platform accessory 'Rökdetektor (Stue)'...
[2018-5-29 19:38:20] [Verisure] Initializing platform accessory 'MOUSE1 (Loft)'...
[2018-5-29 19:38:20] [Verisure] Device MOUSE1 is not yet supported

/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:145
    if (existing.UUID === service.UUID) {
                                  ^
TypeError: Cannot read property 'UUID' of null
    at Accessory.addService (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:145:35)
    at /usr/local/lib/node_modules/homebridge/lib/server.js:432:19
    at Array.forEach (<anonymous>)
    at Server._createAccessory (/usr/local/lib/node_modules/homebridge/lib/server.js:412:14)
    at Server.<anonymous> (/usr/local/lib/node_modules/homebridge/lib/server.js:374:32)
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/util/once.js:16:19
    at /homebridge/node_modules/homebridge-verisure/index.js:126:9
    at /homebridge/node_modules/verisure/index.js:79:7
    at Request._callback (/homebridge/node_modules/verisure/index.js:27:17)
    at Request.self.callback (/homebridge/node_modules/request/request.js:186:22)
    at emitTwo (events.js:126:13)
    at Request.emit (events.js:214:7)
    at Request.<anonymous> (/homebridge/node_modules/request/request.js:1163:10)
    at emitOne (events.js:116:13)
    at Request.emit (events.js:211:7)
    at IncomingMessage.<anonymous> (/homebridge/node_modules/request/request.js:1085:12)

┌───────────────────────────────────────────────────┐
│              npm update check failed              │
│        Try running with sudo or get access        │
│       to the local update config store via        │
│ sudo chown -R $USER:$(id -gn $USER) /root/.config │

**Comment by ptz0n** [comment]
@bjornand1 I can see that you are getting this error output:

> Device MOUSE1 is not yet supported

Looks like you are running the current version from master, v1.6. Currently, only the development version (`dev`) is supporting alarm, contact sensors and door lock. Please see installation instructions at the top of this PR.

**Comment by bjornand1** [comment]
OK, slowly learning here, there is a way to go into command-line mode, but even then v 1.6 is installed when I run 

npm install -g git://github.com/ptz0n/homebridge-verisure#dev

How to make 1.7 install?

**Comment by ptz0n** [comment]
@bjornand1 Oh, I see. I've now published the prerelease `1.7.0-rc.0`, can you see if you can install that one?

**Comment by hoefs** [comment]
@ptz0n & @bjornand1 
I have tested the new version and all seems to work:

pi@retropie:~ $ sudo npm list -g --depth=0
/usr/lib
├── homebridge@0.4.42
├── homebridge-raspberrypi-temperature@0.0.6
├── homebridge-unifi-occupancy-sensor@0.1.1
├── homebridge-verisure@1.6.0 (git://github.com/ptz0n/homebridge-verisure.git#fc058fa22b78a1bd3f0fd12c37c11d79f526598e)
├── homebridge-zp@0.2.20
├── npm@6.1.0
├── npm-check@5.7.1
└── npm-check-updates@2.14.2

pi@retropie:~ $ npm install -g git://github.com/ptz0n/homebridge-verisure#dev
+ homebridge-verisure@1.7.0-rc.0
updated 1 package in 12.487s
pi@retropie:~ $ sudo npm list -g --depth=0
/usr/lib
├── homebridge@0.4.42
├── homebridge-raspberrypi-temperature@0.0.6
├── homebridge-unifi-occupancy-sensor@0.1.1
├── homebridge-verisure@1.7.0-rc.0 (git://github.com/ptz0n/homebridge-verisure.git#956c3fc47abda3e52b04fff6da235f208133a133)
├── homebridge-zp@0.2.20
├── npm@6.1.0
├── npm-check@5.7.1
└── npm-check-updates@2.14.2

**Comment by bjornand1** [comment]
Great, now it works! And much more component functionality than what I had before. Thanks for your effort on this!

**Comment by bjornand1** [comment]
Hi, could I ask, when I updated earlier to a new version of the Verisure plugin, all my Verisure devices were restored to their original Verisure names and put into the default room. I have therefore not updated to 1.9.1 as it takes some work to rename and relocate components. Do you know if that will happen every time the plugin is updated?

BR,
Bjorn

**Comment by ptz0n** [comment]
@bjornand1 I've experienced the same "problem" when updating & developing. I think this might be related to how homebridge is persisting data and seem to be out of our control as plugin developers.

If there is any known way to help homebridge to better keep track of our devices, I'm happy to solve this quite frustrating but not critical issue.

**Comment by bjornand1** [comment]
Ah, so it will definitely happen again if I update, then? I’ll have to weigh the pros of updating against the cons of the extra effort, then. Would be great to find a way to avoid this (I seem to remember that for some previous updates, the names/rooms stuck?).

**Pull request merged**